sql: lift lookup join parallelization into execbuilder

yuzefovich opened the pull request:
In preparation for expanding the heuristic for when it's safe to parallelize cross-range reads in the lookup join, this commit lifts the decision into the execbuilder. As a reminder, when Streamer API is not used, we currently choose to enable parallelism when an input row can result in at most one lookup row (which is the case for index joins as well as lookup joins when equality columns form a key), also when the session variable enables it for multi-lookup case. Previously, this was determined when creating the JoinReader, and now we make this determination in the execbuilder. We add a new field to the processor spec based on the execbuilder argument (for compatibility with 25.2 release we also keep the old logic for now).

Additionally, this commit enhances the EXPLAIN output to show `parallel` attribute in VERBOSE mode when applicable (when the streamer is used as well as when Parallelize is set to true by the execbuilder). This improvement was part of the reason to lifting the determination up there (so that `parallelize` boolean argument to ConstructLookupJoin is captured by the explain factory).

Informs: #134351.
Epic: None

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/148306)
<!-- Reviewable:end -->

yuzefovich committed:
sql: lift lookup join parallelization into execbuilder

In preparation for expanding the heuristic for when it's safe to
parallelize cross-range reads in the lookup join, this commit lifts the
decision into the execbuilder. As a reminder, when Streamer API is not
used, we currently choose to enable parallelism when an input row can
result in at most one lookup row (which is the case for index joins as
well as lookup joins when equality columns form a key), also when the
session variable enables it for multi-lookup case. Previously, this was
determined when creating the JoinReader, and now we make this
determination in the execbuilder. We add a new field to the processor
spec based on the execbuilder argument (for compatibility with 25.2
release we also keep the old logic for now).

Additionally, this commit enhances the EXPLAIN output to show `parallel`
attribute in VERBOSE mode when applicable (when the streamer is used as
well as when Parallelize is set to true by the execbuilder). This
improvement was part of the reason to lifting the determination up there
(so that `parallelize` boolean argument to ConstructLookupJoin is
captured by the explain factory).

Release note: None

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice!

> Reviewed 34 of 34 files at r2, 40 of 41 files at r3, 5 of 5 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/148306#-:-OT0ET8UCxyKo92sWRZ_:b-8gk8uj)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+
